python/demo/demo_navier_stokes.py

```diff
@@ -0,0 +1,419 @@
+                        + alpha / avg(h) * inner(jump(u, n), jump(v, n)) * dS
+                        - inner(grad(u), outer(v, n)) * ds
+                        - inner(outer(u, n), grad(v)) * ds
+                        + alpha / h * inner(outer(u, n), outer(v, n)) * ds)
+a_01 = - inner(p, div(v)) * dx
+a_10 = - inner(div(u), q) * dx
+a_11 = fem.Constant(msh, PETSc.ScalarType(0.0)) * inner(p, q) * dx
+
+a = fem.form([[a_00, a_01],
+              [a_10, a_11]])
+
+f = fem.Function(W)
+u_D = fem.Function(V)
+u_D.interpolate(u_e_expr)
+L_0 = inner(f, v) * dx + \
+    1 / R_e_const * (- inner(outer(u_D, n), grad(v)) * ds
+                     + alpha / h * inner(outer(u_D, n), outer(v, n)) * ds)
+L_1 = inner(fem.Constant(msh, PETSc.ScalarType(0.0)), q) * dx
+
+L = fem.form([L_0,
+              L_1])
+
+# Boundary conditions
+
+boundary_facets = mesh.locate_entities_boundary(
+    msh, msh.topology.dim - 1, boundary_marker)
+boundary_vel_dofs = fem.locate_dofs_topological(
+    V, msh.topology.dim - 1, boundary_facets)
+bc_u = fem.dirichletbc(u_D, boundary_vel_dofs)
+
+# The pressure is only determined up to a constant, so pin a single degree
+# of freedom
+
+# TODO TIDY
+pressure_dofs = fem.locate_dofs_geometrical(
+    Q, lambda x: np.logical_and(np.isclose(x[0], 0.0),
+                                np.isclose(x[1], 0.0)))
+if len(pressure_dofs) > 0:
+    pressure_dof = [pressure_dofs[0]]
+else:
+    pressure_dof = []
+bc_p = fem.dirichletbc(PETSc.ScalarType(0.0),
+                       np.array(pressure_dof, dtype=np.int32),
+                       Q)
+
+bcs = [bc_u, bc_p]
+
+# Assemble Stokes problem
+
+A = fem.petsc.assemble_matrix_block(a, bcs=bcs)
+A.assemble()
+b = fem.petsc.assemble_vector_block(L, a, bcs=bcs)
+
+# Create and configure solver
+
+ksp = PETSc.KSP().create(msh.comm)
+ksp.setOperators(A)
+ksp.setType("preonly")
+ksp.getPC().setType("lu")
+ksp.getPC().setFactorSolverType("mumps")
+opts = PETSc.Options()
+# See https://graal.ens-lyon.fr/MUMPS/doc/userguide_5.5.1.pdf
+# TODO Check
+opts["mat_mumps_icntl_6"] = 2
+opts["mat_mumps_icntl_14"] = 100
+opts["ksp_error_if_not_converged"] = 1
+ksp.setFromOptions()
+
+# Solve Stokes for initial condition
+
+x = A.createVecRight()
+ksp.solve(b, x)
+
+# Split the solution
+
+u_h = fem.Function(V)
+p_h = fem.Function(Q)
+p_h.name = "p"
+offset = V.dofmap.index_map.size_local * V.dofmap.index_map_bs
+u_h.x.array[:offset] = x.array_r[:offset]
+u_h.x.scatter_forward()
+p_h.x.array[:(len(x.array_r) - offset)] = x.array_r[offset:]
+p_h.x.scatter_forward()
+
+u_vis = fem.Function(W)
+u_vis.name = "u"
+u_vis.interpolate(u_h)
+
+# Write initial condition to file
+
+u_file = io.VTXWriter(msh.comm, "u.bp", [u_vis._cpp_object])
+p_file = io.VTXWriter(msh.comm, "p.bp", [p_h._cpp_object])
+t = 0.0
+u_file.write(t)
+p_file.write(t)
+
+# Create function to store solution and previous time step
+
+u_n = fem.Function(V)
+u_n.x.array[:] = u_h.x.array
```

Comment:
```suggestion
u_n.x.array[:] = u_h.x.array_r
```

Reply:
`u_h.x` returns a dolfinx vector, not a PETSc vector